python/ray/data/impl/plan.py

```diff
@@ -44,18 +45,18 @@ def with_stage(self, stage: "Stage") -> "ExecutionPlan":
         Returns:
             A new ExecutionPlan with this stage appended.
         """
+        copy = ExecutionPlan(self._in_blocks, self._in_stats)
         if self._out_blocks:
-            copy = ExecutionPlan(self._out_blocks, self._out_stats)
-            copy._stages = [stage]
-        else:
-            copy = ExecutionPlan(self._in_blocks, self._in_stats)
-            copy._stages = self._stages.copy()
-            copy._stages.append(stage)
+            copy._out_blocks = self._out_blocks
+            copy._out_stats = self._out_stats
+            copy._out_stage = self._out_stage
+        copy._stages = self._stages.copy()
+        copy._stages.append(stage)
```

Comment:
Note that we preserve the full Dataset lineage now instead of tracking lineage only to the last `_output_block`. Since we need to serialize the full lineage back to the datasource, this change is required.